src/Model/GroundWaterFlow/gwf3vsc8.f90

```diff
@@ -0,0 +1,1519 @@
+    ! -- return
+    real(DP) :: viscbnd
+    ! -- local
+    integer(I4B) :: i
+! ------------------------------------------------------------------------------
+    !
+    ! -- assign boundary viscosity based on one of three options
+    if (locvisc > 0) then
+      ! -- assign viscosity to an aux column named 'VISCOSITY'
+      viscbnd = auxvar(locvisc, n)
+    else if (locconc(1) > 0) then
+      ! -- calculate viscosity using one or more concentration auxcolumns
+      do i = 1, size(locconc)
+        ctemp(i) = DZERO
+        if (locconc(i) > 0) then
+          ctemp(i) = auxvar(locconc(i), n)
+        end if
+      end do
+      viscbnd = calc_visc(ivisc, viscref, dviscdc, cviscref, ctemp, a2, a3, a4)
+    else
+      ! -- neither of the above, so assign as viscref
+      viscbnd = viscref
+    end if
+    !
+    ! -- return
+    return
+  end function calc_bnd_viscosity
+
+  !> @brief Account for viscosity in horizontal flow barriers
+  !!
+  !! Will return the viscosity associated with the upgradient node (cell)
+  !! to the HFB package for adjusting the hydraulic characteristic (hydchr)
+  !! of the barrier
+  !<
+  subroutine set_hfb_visc(this, cellid, hfbviscratio)
+    ! -- dummy variables
+    class(GwfVscType) :: this
+    integer(I4B), intent(in) :: cellid
+    ! -- return
+    real(DP), intent(inout) :: hfbviscratio
+    ! -- local
+    real(DP) :: visc
+! ------------------------------------------------------------------------------
+    !
+    ! -- Retrieve viscosity for the passed node number
+    visc = this%visc(cellid)
+    !
+    ! -- Calculate the viscosity ratio for the
+    hfbviscratio = calc_vsc_ratio(this%viscref, visc)
+    !
+    ! -- return
+    return
+  end subroutine set_hfb_visc
+
+  !> @brief hit the hydraulic conductivity values with the ratio mu_o/mu
+  !!
+  !! This routine called after updating the viscosity values using the latest
+  !! concentration and/or temperature values.  The ratio mu_o/mu, reference
+  !! viscosity divided by the updated viscosity value, is multiplied by K
+  !! for each cell.
+  !<
+  subroutine update_k_with_vsc(this)
+!
+!    SPECIFICATIONS:
+! ------------------------------------------------------------------------------
+    ! -- modules
+    ! -- dummy
+    class(GwfVscType) :: this
+    integer(I4B) :: n
+! ------------------------------------------------------------------------------
+    !
+    ! -- For viscosity-based K's, apply change of K to K11 by starting with
+    !    user-specified K values and not the K's leftover from the last viscosity
+    !    update.
+    do n = 1, this%dis%nodes
+      this%k11(n) = this%k11input(n) * (this%viscref / this%visc(n))
+      if (this%ik22 /= 0) then
+        this%k22(n) = this%k22input(n) * (this%viscref / this%visc(n))
+      end if
+      if (this%ik33 /= 0) then
+        this%k33(n) = this%k33input(n) * (this%viscref / this%visc(n))
+      end if
+      this%nodekchange(n) = 1
```

Comment:
To leverage existing conductance calculations within NPF, the `k11`, `k22`, and `k33` arrays are multiplied by the viscosity ratio early on so that all the subsequent uses of `k11` work as currently coded (no need for new code) and account for temperature/concentration.  By taking this approach, the `k` arrays get stomped on by VSC, meaning the original user-specified values are lost...unless they are backed up in `k11input`, for example.  So in the code highlighted by this comment,

`this%k11(n) = this%k11input(n) * (this%viscref / this%visc(n))`

`k11input(n)` is the safe, user-specified value.  `k11` may reflect a modified value from a previous stress period and therefore is not the correct value to update while accounting for the effects of viscosity.  All this to say, I think this logic still applies.

I think the fix reflected in #1031 is separate from what's going on here.

Reply:
In NPF, I think we now use K11, K22, and K33 in the calculations for conductance (in npf and in xt3d).  We use these arrays even if ik22 = 0 (e.g., the user did not specify K22).  It looks to me like you are only updating k22 if the user specifies k22 in the input files, but to me that suggests K22 is not updated with viscosity effects when it should be.  But I could certainly be misunderstanding something.

Reply:
Ah, I mis-understood the first comment:
> Since we are always allocating and storing K11, K22, and K33, even if not specified by the user, does this logic still apply?
Indeed a correction is needed here.  The if statements surrounding `k22` and `k33` should be removed if both are in regular use.